### PR TITLE
mips: debug: add stack iterator

### PR DESCRIPTION
**Pull request by nastya-nizharadze**

### CODE REVIEW

File: src/arch/mips/lib/debug/stack_iter.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+int stack_iter_next(stack_iter_t *f) {
+	uint32_t offset_cur;
+
+	if (!(offset_cur = offset(f->ra))) {
+		return 0;
+	}
+/* To get an old fp we need to cast the contents of this  |
+ * block of memory to the void pointer type				  V */
```
Comment (alexkalmuk) [inline src/arch/mips/lib/debug/stack_iter.c]: `type <many blanks> V` something wrong with formatting here?

File: src/arch/mips/lib/debug/stack_iter.c [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	f->fp = (void *) *((unsigned *) (f->fp + offset_cur - 8));
+	f->ra = (void *) *((unsigned *) (f->fp - 4));
+/* When the old fp become 0 stack is ower */
+	return (void *) *((unsigned *) (f->fp + offset_cur - 8)) != 0;
```
Comment (alexkalmuk) [inline src/arch/mips/lib/debug/stack_iter.c]: What happens here? You already updated `f->fp` to the new value above:

```
f->fp = (void *) *((unsigned *) (f->fp + offset_cur - 8));
```

And at this line you increment `fp` at the same value once more in order to compare with 0.

File: src/arch/mips/lib/debug/stack_iter.c (hunk 1 above)
Comment (nastya-nizharadze) [reply src/arch/mips/lib/debug/stack_iter.c]: done

File: src/arch/mips/lib/debug/stack_iter.c [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * | old fp |  |		<- new fp + offset - 8
+ * |   ...  |  | offset
+ * |        |  |
+ * |        |  v
+ * |        | ---		<- new fp
+ *  -------- frame ending
+ * 
+ *  */
+
+int offset (void *ra) {
```
Comment (Deryugin) [inline src/arch/mips/lib/debug/stack_iter.c]: Is this function used outside of this file? If no, make it `static`, please

File: src/arch/mips/lib/debug/stack_iter.c (hunk 2 above)
Comment (nastya-nizharadze) [reply src/arch/mips/lib/debug/stack_iter.c]: It is supposed to be a next fp. Done.

File: src/arch/mips/lib/debug/stack_iter.c (hunk 3 above)
Comment (Deryugin) [reply src/arch/mips/lib/debug/stack_iter.c]: `offset (`->`offset(`

Comment (nastya-nizharadze) [reply src/arch/mips/lib/debug/stack_iter.c]: done

**Comment by alexkalmuk** [comment]
@nastya-nizharadze The code looks good to me! But stack tracing doesn't work as expected when assertion happen (we use stack tracing mostly for assertions and faults tracing). I inserted `assert(0)` in src/cmds/help.c:main, build Embox and launch QEMU. After typing help I get the following:

```
 ASSERTION FAILED on CPU 0
	at src/cmds/help.c:25
	in function main

0


 --   00001000 * A R    thread 15  task 4 -------------------------------------


run 0x80006fd0 <task_trampoline+0x0>                 src/kernel/task/multi.c:47
```

On x86/qemu I got for the same patсh:

```
 ASSERTION FAILED on CPU 0
	at src/cmds/help.c:25
	in function main

0


 --   00001000 * A R    thread 14  task 4 -------------------------------------

  6 0x0014649a <__assertion_handle_failure+0x114> ...at/libc/assert/assert.c:45
  5 0x0012fbe2 <__module_embox__cmd__man_text_vma+0x0> ...ramework/cmd/api.h:34
  4 0x00181cb7 <cmd_exec+0x5c>                      src/framework/cmd/core.c:23
  3 0x0011db13 <run_cmd+0x83>                         src/cmds/shell/tish.c:159
  2 0x00109808 <task_trampoline+0x27>                src/kernel/task/multi.c:47
  1 0x0010a81c <thread_trampoline+0xa8>             src/kernel/thread/core.c:69

run 0x001097e1 <task_trampoline+0x0>                 src/kernel/task/multi.c:47
```

So as you can see on qemu mips we miss the whole stack trace somewhy. Can I ask you to take a look at this, please?

File: src/arch/mips/lib/debug/stack_iter.c [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	if (a == NULL) {
+		return 0;
+	}
+/* Find an operation code. it is always addiu sp,sp,x 
+ * where x is an offset */
+	opcode = *((unsigned *) (a->addr));
+/* The operation code for addiu sp,sp,x should always started whith 0X27bd */
+	assert((opcode & 0xffff0000) == 0x27bd0000);
+/* The last 4 digits are for offset*/
+	return (0x10000 - (opcode & 0xffff));
```
Comment (alexkalmuk) [inline src/arch/mips/lib/debug/stack_iter.c]: I'd propose to specify directly that this function assumes offset _must_ be negative.

```
int16_t offset;
offset = opcode & 0xffff;
assert(offset < 0);
return -offset;
```

Comment (nastya-nizharadze) [reply src/arch/mips/lib/debug/stack_iter.c]: done

**Pull request merged**